client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java

```diff
@@ -1586,4 +1708,134 @@ private boolean connectFail(String message) {
   public TransportClientFactory getDataClientFactory() {
     return dataClientFactory;
   }
+
+  class ReviveRequest {
+    public int mapId;
+    public int attemptId;
+    public int partitionId;
+    public PartitionLocation loc;
+    public StatusCode cause;
+    public volatile StatusCode reviveStatus;
+
+    public ReviveRequest(
+        int mapId, int attemptId, int partitionId, PartitionLocation loc, StatusCode cause) {
+      this.mapId = mapId;
+      this.attemptId = attemptId;
+      this.partitionId = partitionId;
+      this.loc = loc;
+      this.cause = cause;
+      reviveStatus = StatusCode.REVIVE_INITIALIZED;
+    }
+  }
+
+  class ReviveManager {
+    // appid -> shuffleId -> requests
+    HashMap<String, HashMap<Integer, Set<ReviveRequest>>> pendingRequests = new HashMap<>();
+    HashMap<String, HashMap<Integer, Set<ReviveRequest>>> inProcessRequests = null;
+    private ScheduledExecutorService batchReviveRequestScheduler =
+        ThreadUtils.newDaemonSingleThreadScheduledExecutor("worker-forward-message-scheduler");
+
+    public ReviveManager() {
+      batchReviveRequestScheduler.scheduleAtFixedRate(
+          new Runnable() {
+            @Override
+            public void run() {
+              // We must ensure the swap is atomic
+              synchronized (this) {
```

Comment:
This synchronization block can be replaced with ReentrantReadWriteLock's WriteLock.

Reply:
There is another drawback of `synchronized`, it does not work well with [Loom](https://openjdk.org/jeps/425), not sure if we have chance to use it in the future

Reply:
> This synchronization block can be replaced with ReentrantReadWriteLock's WriteLock.

ditto

Reply:
> This

Thanks @pan3793 for introducing Java virtual threads. It seems very like coroutine. I think we can explore to find out which cases Celeborn can leverage it. In my opinion it's best suitable for non-collaborative, high throughput with IO-blocking operations scenarios. It's a better programing pattern than current future-callback pattern.

However, collaborative logic is unavoidable and so is critical section. I think in this piece of code the synchronization is unavoidable. Of cause we can re-test the perf of synchronization and RWLock.

Reply:
+1 for `synchronized`, just share the related context since we are talking about this area.

Reply:
> +1 for `synchronized`, just share the related context since we are talking about this area.

Thanks :)

Reply:
Hi @FMX , could you please paste your test on synchronized and RWLock here?